src/dwt.rs

```diff
@@ -0,0 +1,219 @@
+        } else if ticks <= u32::MAX as u64 {
+            // Try to avoid race conditions by limiting delay to u32::MAX / 2
+            let mut ticks = ticks as u32;
+            ticks -= u32::MAX / 2;
+            while (DWT::get_cycle_count().wrapping_sub(start)) < u32::MAX / 2 {}
+            start -= u32::MAX / 2;
+            while (DWT::get_cycle_count().wrapping_sub(start)) < ticks {}
+        } else {
+            // Delay for ticks, then delay for rest * u32::MAX
+            let mut rest = (ticks >> 32) as u32;
+            let ticks = (ticks & u32::MAX as u64) as u32;
+            loop {
+                while (DWT::get_cycle_count().wrapping_sub(start)) < ticks {}
+                if rest == 0 {
+                    break;
+                }
+                rest -= 1;
+                while (DWT::get_cycle_count().wrapping_sub(start)) > ticks {}
+            }
+        }
+    }
+}
+
+// Implement DelayUs/DelayMs for various integer types
+impl DelayUs<u64> for Delay {
+    fn delay_us(&mut self, us: u64) {
+        // Convert us to ticks
+        let start = DWT::get_cycle_count();
+        let ticks = (us * self.clock.0 as u64) / 1_000_000;
+        Delay::delay_ticks(start, ticks);
+    }
+}
+impl DelayMs<u64> for Delay {
+    fn delay_ms(&mut self, ms: u64) {
+        // Convert ms to ticks
+        let start = DWT::get_cycle_count();
+        let ticks = (ms * self.clock.0 as u64) / 1_000;
+        Delay::delay_ticks(start, ticks);
+    }
+}
+macro_rules! impl_DelayIntT {
+    (for $($t:ty),+) => {$(
+        impl DelayUs<$t> for Delay {
+            fn delay_us(&mut self, us: $t) {
+                self.delay_us(us as u64);
+            }
+        }
+        impl DelayMs<$t> for Delay {
+            fn delay_ms(&mut self, ms: $t) {
+                self.delay_ms(ms as u64);
+            }
+        }
+    )*}
+}
+impl_DelayIntT!(for usize,  u32, u16, u8);
+
+/// Very simple stopwatch
+pub struct StopWatch<'l> {
+    times: &'l mut [u32],
+    timei: usize,
+    clock: Hertz,
+}
+impl<'l> StopWatch<'l> {
+    /// Create a new instance (Private because dwt/dcb should be set up)
+    /// # Arguments
+    /// * `times` - Array which will be holding the timings (max laps == times.len()-1)
+    /// * `clock` - The DWT cycle counters clock
+    fn new(times: &'l mut [u32], clock: Hertz) -> Self {
+        assert!(times.len() >= 2);
+        let mut sw = StopWatch {
+            times,
+            timei: 0,
+            clock,
+        };
+        sw.reset();
+        sw
+    }
+    /// Returns the numbers of laps recorded
+    pub fn lap_count(&self) -> usize {
+        self.timei
+    }
+    /// Resets recorded laps to 0 and sets 0 offset
+    pub fn reset(&mut self) {
+        self.timei = 0;
+        self.times[0] = DWT::get_cycle_count();
+    }
+    /// Record a new lap
+    /// NOTE If lap count exceeds maximum, the last lap is updated
+    pub fn lap(&mut self) -> &Self {
```

Comment:
Why returning &Self (instead of nothing or &must Self)?

Reply:
I return nothing here, so I can write `sw.lap().lap_time(1).unwrap()` (see `fn measure(..)`).
Fixed the return type to `&mut Self`.